File: webapp/src/compiler/parser.ts
import { TokenType } from './tokens';
import type { TokenTypeId } from './tokens';
import type { Token } from './tokens';
import { tokenize } from './lexer';
import type { Program, Line, Stmt, Expr } from './ast';

export class ParseError extends Error {
  token: Token | null;
  constructor(message: string, token: Token | null = null) {
    super(token ? `${message} at line ${token.line}, column ${token.column}` : message);
    this.name = 'ParseError';
    this.token = token;
  }
}

export function parse(source: string): Program {
  const tokens = tokenize(source);
  const p = new Parser(tokens);
  return p.parse();
}

class Parser {
  pos = 0;
  tokens: Token[];
  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  current(): Token {
    return this.pos >= this.tokens.length ? this.tokens[this.tokens.length - 1] : this.tokens[this.pos];
  }
  isType(tt: TokenTypeId): boolean {
    return this.current().type === tt;
  }
  consume(tt: TokenTypeId, msg = ''): Token {
    if (this.current().type !== tt) throw new ParseError(msg || `Expected ${tt}`, this.current());
    const t = this.current();
    this.pos++;
    return t;
  }
  consumeIf(tt: TokenTypeId): boolean {
    if (this.current().type === tt) { this.pos++; return true; }
    return false;
  }
  skipNewlines(): void {
    while (this.consumeIf(TokenType.NEWLINE)) {}
  }

  parse(): Program {
    const lines: Line[] = [];
    this.skipNewlines();
    while (this.current().type !== TokenType.EOF) {
      const line = this.parseLine();
      if (line) lines.push(line);
      this.skipNewlines();
    }
    return { lines };
  }

  parseLine(): Line | null {
    let lineNum: number | null = null;
    if (this.isType(TokenType.NUMBER)) {
      lineNum = Number(this.current().value);
      this.pos++;
    }
    const statements: Stmt[] = [];
    let stmt = this.parseStatement();
    if (stmt) statements.push(stmt);
    while (this.consumeIf(TokenType.COLON)) {
      stmt = this.parseStatement();
      if (stmt) statements.push(stmt);
    }
    if (lineNum !== null || statements.length > 0) return { number: lineNum, statements };
    return null;
  }

  parseStatement(): Stmt | null {
    const t = this.current();
    if (t.type === TokenType.PRINT) return this.parsePrint();
    if (t.type === TokenType.LET) return this.parseLet();
    if (t.type === TokenType.INPUT) return this.parseInput();
    if (t.type === TokenType.IF) return this.parseIf();
    if (t.type === TokenType.GOTO) return this.parseGoto();
    if (t.type === TokenType.GOSUB) return this.parseGosub();
    if (t.type === TokenType.RETURN) return this.parseReturn();
    if (t.type === TokenType.FOR) return this.parseFor();
    if (t.type === TokenType.NEXT) return this.parseNext();
    if (t.type === TokenType.END) return this.parseEnd();
    if (t.type === TokenType.REM) return this.parseRem();
    if (t.type === TokenType.NEWLINE || t.type === TokenType.COLON || t.type === TokenType.EOF) return null;
    throw new ParseError(`Unexpected token: ${t.type}`, t);
  }

  parsePrint(): Stmt {
    this.consume(TokenType.PRINT);
    const items: Expr[] = [];
    for (;;) {
      if (this.isType(TokenType.STRING)) {
        items.push({ kind: 'string', value: String(this.current().value) });
        this.pos++;
      } else {
        items.push(this.parseExpression());
      }
      if (!this.consumeIf(TokenType.COMMA)) break;
    }
    return { kind: 'print', items };
  }
  parseLet(): Stmt {
    this.consume(TokenType.LET);
    if (!this.isType(TokenType.IDENT)) throw new ParseError('Expected variable name', this.current());
    const name = String(this.current().value).toUpperCase();
    this.pos++;
    this.consume(TokenType.EQ, 'Expected =');
    const value = this.parseExpression();
    return { kind: 'let', name, value };
  }
  parseInput(): Stmt {
    this.consume(TokenType.INPUT);
    const variables: string[] = [];
    for (;;) {
      if (!this.isType(TokenType.IDENT)) throw new ParseError('Expected variable name', this.current());
      variables.push(String(this.current().value).toUpperCase());
      this.pos++;
      if (!this.consumeIf(TokenType.COMMA)) break;
    }
    return { kind: 'input', variables };
  }
  parseRelop(): string {
    const t = this.current();
    if (t.type === TokenType.LT) { this.pos++; return '<'; }
    if (t.type === TokenType.LE) { this.pos++; return '<='; }
    if (t.type === TokenType.GT) { this.pos++; return '>'; }
    if (t.type === TokenType.GE) { this.pos++; return '>='; }
    if (t.type === TokenType.EQ) { this.pos++; return '='; }
    if (t.type === TokenType.NE) { this.pos++; return '<>'; }
    throw new ParseError('Expected comparison operator', t);
  }
  parseIf(): Stmt {
    this.consume(TokenType.IF);
    const left = this.parseExpression();
    const relop = this.parseRelop();
    const right = this.parseExpression();
    this.consume(TokenType.THEN, 'Expected THEN');
    const then_stmt = this.parseStatement();
    if (!then_stmt) throw new ParseError('Expected statement after THEN', this.current());
    let else_stmt: Stmt | undefined;
    if (this.consumeIf(TokenType.ELSE)) else_stmt = this.parseStatement() ?? undefined;
    return { kind: 'if', left, relop, right, then_stmt, else_stmt };
  }
  parseGoto(): Stmt {
    this.consume(TokenType.GOTO);
    return { kind: 'goto', target: this.parseExpression() };
  }
  parseGosub(): Stmt {
    this.consume(TokenType.GOSUB);
    return { kind: 'gosub', target: this.parseExpression() };
  }
  parseReturn(): Stmt {
    this.consume(TokenType.RETURN);
    return { kind: 'return' };
  }
  parseFor(): Stmt {
    this.consume(TokenType.FOR);
    if (!this.isType(TokenType.IDENT)) throw new ParseError('Expected variable in FOR', this.current());
    const varName = String(this.current().value).toUpperCase();
    this.pos++;
    this.consume(TokenType.EQ, 'Expected = in FOR');
    const start = this.parseExpression();
    this.consume(TokenType.TO, 'Expected TO in FOR');
    const end = this.parseExpression();
    let step: Expr | undefined;
    if (this.consumeIf(TokenType.STEP)) step = this.parseExpression();
    const body: Stmt[] = [];
    for (;;) {
      if (this.isType(TokenType.NEXT)) {
        this.pos++;
        if (this.isType(TokenType.IDENT)) this.pos++;
        break;
      }
      if (this.isType(TokenType.EOF)) throw new ParseError('FOR without NEXT', this.current());
      if (this.isType(TokenType.NEWLINE)) { this.skipNewlines(); continue; }
      if (this.isType(TokenType.COLON)) { this.pos++; continue; }
      if (this.isType(TokenType.NUMBER)) { this.pos++; continue; }
      const stmt = this.parseStatement();
      if (stmt) body.push(stmt);
    }
    return { kind: 'for', var: varName, start, end, step, body };
  }
  parseNext(): Stmt {
    this.consume(TokenType.NEXT);
    let varName: string | undefined;
    if (this.isType(TokenType.IDENT)) { varName = String(this.current().value).toUpperCase(); this.pos++; }
    return { kind: 'next', var: varName };
  }
  parseEnd(): Stmt {
    this.consume(TokenType.END);
    return { kind: 'end' };
  }
  parseRem(): Stmt {
    const t = this.consume(TokenType.REM);
    return { kind: 'rem', text: String(t.value ?? '') };
  }
  parseExpression(): Expr {
    if (this.consumeIf(TokenType.PLUS)) return this.parseTerm();
    if (this.consumeIf(TokenType.MINUS)) return { kind: 'unary', op: '-', operand: this.parseTerm() };
    let expr = this.parseTerm();
    for (;;) {
      if (this.consumeIf(TokenType.PLUS)) expr = { kind: 'binary', op: '+', left: expr, right: this.parseTerm() };
      else if (this.consumeIf(TokenType.MINUS)) expr = { kind: 'binary', op: '-', left: expr, right: this.parseTerm() };
      else break;
    }
    return expr;
  }
  parseTerm(): Expr {
    let expr = this.parseFactor();
    for (;;) {
      if (this.consumeIf(TokenType.STAR)) expr = { kind: 'binary', op: '*', left: expr, right: this.parseFactor() };
      else if (this.consumeIf(TokenType.SLASH)) expr = { kind: 'binary', op: '/', left: expr, right: this.parseFactor() };
      else break;
    }
    return expr;
  }
  parseFactor(): Expr {
    const t = this.current();
    if (t.type === TokenType.NUMBER) { this.pos++; return { kind: 'number', value: Number(t.value) }; }
    if (t.type === TokenType.STRING) { this.pos++; return { kind: 'string', value: String(t.value) }; }
    if (t.type === TokenType.IDENT) {
      const name = String(t.value).toUpperCase();
      this.pos++;
      if (this.isType(TokenType.LPAREN)) {
        this.pos++;
        const args: Expr[] = [];
        if (!this.isType(TokenType.RPAREN)) {
          args.push(this.parseExpression());
          while (this.consumeIf(TokenType.COMMA)) args.push(this.parseExpression());
        }
        this.consume(TokenType.RPAREN, 'Expected )');
        return { kind: 'builtin', name, args };
      }
      return { kind: 'var', name };
    }
    if (this.consumeIf(TokenType.LPAREN)) {
      const expr = this.parseExpression();
      this.consume(TokenType.RPAREN, 'Expected )');
      return expr;
    }
    throw new ParseError('Expected expression', t);
  }
}

File: src/parser.py
"""
Recursive descent parser for BASIC. Produces AST (Program with Lines and Stmts).
"""
from typing import List, Optional

from .tokens import Token, TokenType
from .ast_nodes import (
    Program, Line, Stmt, Expr,
    PrintStmt, LetStmt, InputStmt, IfStmt, GotoStmt, GosubStmt, ReturnStmt,
    ForStmt, NextStmt, EndStmt, RemStmt,
    NumberExpr, StringExpr, VarExpr, BinaryOpExpr, UnaryOpExpr, BuiltinCallExpr,
)


class ParseError(Exception):
    def __init__(self, message: str, token: Optional[Token] = None):
        self.message = message
        self.token = token
        if token:
            super().__init__(f"{message} at line {token.line}, column {token.column}")
        else:
            super().__init__(message)


class Parser:
    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.pos = 0

    def _current(self) -> Token:
        if self.pos >= len(self.tokens):
            return self.tokens[-1]
        return self.tokens[self.pos]

    def _is_type(self, tt: TokenType) -> bool:
        return self._current().type == tt

    def _consume(self, tt: TokenType, msg: str = "") -> Token:
        if self._current().type != tt:
            raise ParseError(msg or f"Expected {tt.name}", self._current())
        t = self._current()
        self.pos += 1
        return t

    def _consume_if(self, tt: TokenType) -> bool:
        if self._current().type == tt:
            self.pos += 1
            return True
        return False

    def _skip_newlines(self) -> None:
        while self._consume_if(TokenType.NEWLINE):
            pass

    def parse(self) -> Program:
        lines: List[Line] = []
        self._skip_newlines()
        while self._current().type != TokenType.EOF:
            line = self._parse_line()
            if line is not None:
                lines.append(line)
            self._skip_newlines()
        return Program(lines=lines)

    def _parse_line(self) -> Optional[Line]:
        # Optional line number (number at start of line)
        line_num: Optional[int] = None
        if self._is_type(TokenType.NUMBER):
            line_num = int(self._current().value)
            self.pos += 1
        statements: List[Stmt] = []
        # First statement
        stmt = self._parse_statement()
        if stmt is not None:
            statements.append(stmt)
        # More statements after COLON
        while self._consume_if(TokenType.COLON):
            stmt = self._parse_statement()
            if stmt is not None:
                statements.append(stmt)
        # If we had only line number and no statements (or only REM), still add the line
        if line_num is not None or statements:
            return Line(number=line_num, statements=statements)
        return None

    def _parse_statement(self) -> Optional[Stmt]:
        t = self._current()
        if t.type == TokenType.PRINT:
            return self._parse_print()
        if t.type == TokenType.LET:
            return self._parse_let()
        if t.type == TokenType.INPUT:
            return self._parse_input()
        if t.type == TokenType.IF:
            return self._parse_if()
        if t.type == TokenType.GOTO:
            return self._parse_goto()
        if t.type == TokenType.GOSUB:
            return self._parse_gosub()
        if t.type == TokenType.RETURN:
            return self._parse_return()
        if t.type == TokenType.FOR:
            return self._parse_for()
        if t.type == TokenType.NEXT:
            return self._parse_next()
        if t.type == TokenType.END:
            return self._parse_end()
        if t.type == TokenType.REM:
            return self._parse_rem()
        if t.type in (TokenType.NEWLINE, TokenType.COLON, TokenType.EOF):
            return None
        raise ParseError(f"Unexpected token: {t.type.name}", t)

    def _parse_print(self) -> PrintStmt:
        self._consume(TokenType.PRINT)
        items: List[Expr] = []
        while True:
            if self._is_type(TokenType.STRING):
                items.append(StringExpr(value=self._current().value))
                self.pos += 1
            else:
                items.append(self._parse_expression())
            if not self._consume_if(TokenType.COMMA):
                break
        return PrintStmt(items=items)

    def _parse_let(self) -> LetStmt:
        self._consume(TokenType.LET)
        if not self._is_type(TokenType.IDENT):
            raise ParseError("Expected variable name", self._current())
        name = self._current().value.upper()
        self.pos += 1
        self._consume(TokenType.EQ, "Expected =")
        value = self._parse_expression()
        return LetStmt(name=name, value=value)

    def _parse_input(self) -> InputStmt:
        self._consume(TokenType.INPUT)
        variables: List[str] = []
        while True:
            if not self._is_type(TokenType.IDENT):
                raise ParseError("Expected variable name", self._current())
            variables.append(self._current().value.upper())
            self.pos += 1
            if not self._consume_if(TokenType.COMMA):
                break
        return InputStmt(variables=variables)

    def _parse_if(self) -> IfStmt:
        self._consume(TokenType.IF)
        left = self._parse_expression()
        relop = self._parse_relop()
        right = self._parse_expression()
        self._consume(TokenType.THEN, "Expected THEN")
        then_stmt = self._parse_statement()
        if then_stmt is None:
            raise ParseError("Expected statement after THEN", self._current())
        else_stmt = None
        if self._consume_if(TokenType.ELSE):
            else_stmt = self._parse_statement()
        return IfStmt(left=left, relop=relop, right=right, then_stmt=then_stmt, else_stmt=else_stmt)

    def _parse_relop(self) -> str:
        t = self._current()
        if t.type == TokenType.LT:
            self.pos += 1
            return "<"
        if t.type == TokenType.LE:
            self.pos += 1
            return "<="
        if t.type == TokenType.GT:
            self.pos += 1
            return ">"
        if t.type == TokenType.GE:
            self.pos += 1
            return ">="
        if t.type == TokenType.EQ:
            self.pos += 1
            return "="
        if t.type == TokenType.NE:
            self.pos += 1
            return "<>"
        raise ParseError("Expected comparison operator", t)

    def _parse_goto(self) -> GotoStmt:
        self._consume(TokenType.GOTO)
        target = self._parse_expression()
        return GotoStmt(target=target)

    def _parse_gosub(self) -> GosubStmt:
        self._consume(TokenType.GOSUB)
        target = self._parse_expression()
        return GosubStmt(target=target)

    def _parse_return(self) -> ReturnStmt:
        self._consume(TokenType.RETURN)
        return ReturnStmt()

    def _parse_for(self) -> ForStmt:
        self._consume(TokenType.FOR)
        if not self._is_type(TokenType.IDENT):
            raise ParseError("Expected variable in FOR", self._current())
        var = self._current().value.upper()
        self.pos += 1
        self._consume(TokenType.EQ, "Expected = in FOR")
        start = self._parse_expression()
        self._consume(TokenType.TO, "Expected TO in FOR")
        end = self._parse_expression()
        step = None
        if self._consume_if(TokenType.STEP):
            step = self._parse_expression()
        # FOR body: collect statements until NEXT (same line or following lines)
        body: List[Stmt] = []
        while True:
            if self._is_type(TokenType.NEXT):
                self.pos += 1
                if self._is_type(TokenType.IDENT):
                    self.pos += 1  # optional variable after NEXT
                break
            if self._is_type(TokenType.EOF):
                raise ParseError("FOR without NEXT", self._current())
            if self._is_type(TokenType.NEWLINE):
                self._skip_newlines()
                continue
            if self._is_type(TokenType.COLON):
                self.pos += 1
                continue
            if self._is_type(TokenType.NUMBER):
                self.pos += 1  # line number on new line
                continue
            stmt = self._parse_statement()
            if stmt is not None:
                body.append(stmt)
        return ForStmt(var=var, start=start, end=end, step=step, body=body)

    def _parse_next(self) -> NextStmt:
        self._consume(TokenType.NEXT)
        var = None
        if self._is_type(TokenType.IDENT):
            var = self._current().value.upper()
            self.pos += 1
        return NextStmt(var=var)

    def _parse_end(self) -> EndStmt:
        self._consume(TokenType.END)
        return EndStmt()

    def _parse_rem(self) -> RemStmt:
        t = self._consume(TokenType.REM)
        return RemStmt(text=t.value or "")

    def _parse_expression(self) -> Expr:
        # Unary +/-
        if self._consume_if(TokenType.PLUS):
            return self._parse_term()
        if self._consume_if(TokenType.MINUS):
            return UnaryOpExpr(op="-", operand=self._parse_term())
        expr = self._parse_term()
        while True:
            if self._consume_if(TokenType.PLUS):
                expr = BinaryOpExpr(op="+", left=expr, right=self._parse_term())
            elif self._consume_if(TokenType.MINUS):
                expr = BinaryOpExpr(op="-", left=expr, right=self._parse_term())
            else:
                break
        return expr

    def _parse_term(self) -> Expr:
        expr = self._parse_factor()
        while True:
            if self._consume_if(TokenType.STAR):
                expr = BinaryOpExpr(op="*", left=expr, right=self._parse_factor())
            elif self._consume_if(TokenType.SLASH):
                expr = BinaryOpExpr(op="/", left=expr, right=self._parse_factor())
            else:
                break
        return expr

    def _parse_factor(self) -> Expr:
        t = self._current()
        if t.type == TokenType.NUMBER:
            self.pos += 1
            return NumberExpr(value=t.value)
        if t.type == TokenType.STRING:
            self.pos += 1
            return StringExpr(value=t.value)
        if t.type == TokenType.IDENT:
            name = t.value.upper()
            self.pos += 1
            if self._is_type(TokenType.LPAREN):
                self.pos += 1
                args: List[Expr] = []
                if not self._is_type(TokenType.RPAREN):
                    args.append(self._parse_expression())
                    while self._consume_if(TokenType.COMMA):
                        args.append(self._parse_expression())
                self._consume(TokenType.RPAREN, "Expected )")
                return BuiltinCallExpr(name=name, args=args)
            return VarExpr(name=name)
        if self._consume_if(TokenType.LPAREN):
            expr = self._parse_expression()
            self._consume(TokenType.RPAREN, "Expected )")
            return expr
        raise ParseError("Expected expression", t)


def parse(source: str) -> Program:
    from .lexer import tokenize
    tokens = tokenize(source)
    return Parser(tokens).parse()
